### PR TITLE
Support loading TorchSharp models in Python

### PR DESCRIPTION
**Pull request by Leo-LiHao**
See Issue https://github.com/dotnet/TorchSharp/issues/551

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=552) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=552) before we can accept your contribution.<br/><br/>:x: Leo-LiHao [sign now](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=552)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/TorchSharp?pullRequest=552) it.</sub>

File: src/Python/importsd.py
Comment (Leo-LiHao) [inline src/Python/importsd.py:76]: can use `torch.frombuffer` instead, however, it was introduced in version 1.10.x (I am currently using 1.9.x)

File: src/Python/importsd.py
Comment (Leo-LiHao) [inline src/Python/importsd.py:40]: Do not know how to calculate the size of bool and bfloat16

File: src/Python/importsd.py
Comment (Leo-LiHao) [inline src/Python/importsd.py:46]: Do not know the corresponding data type

File: src/Python/importsd.py
Comment (NiklasGustafsson) [reply src/Python/importsd.py:40]: I'm confused -- 'type_num == 0' is your first case in this long if-then-else.

**Comment by NiklasGustafsson** [comment]
@Leo-LiHao -- welcome to our community of contributors!

I would like to ask you to introduce yourself to the community with a paragraph or two in this discussion thread: https://github.com/dotnet/TorchSharp/discussions/333

It's nice for people using the library to know who's been helping put it together.

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]

**Comment by NiklasGustafsson** [comment]
@Leo-LiHao -- I suggested changes to this PR. Have you looked at my comment?

**Comment by Leo-LiHao** [comment]
> @Leo-LiHao -- I suggested changes to this PR. Have you looked at my comment?

Yes, but this will be a busy week for me. I will fix the code & try to add a test this weekend :)

File: test/TorchSharpTest/pyimporttest.py
Comment (NiklasGustafsson) [inline test/TorchSharpTest/pyimporttest.py:33]: @Leo-LiHao-- I think it would be better if there were a test that the loaded model could actually be used in both train and eval mode.

**Comment by NiklasGustafsson** [comment]
@Leo-LiHao -- did you see the build failures?

**Comment by Leo-LiHao** [comment]
> @Leo-LiHao -- did you see the build failures?

Yes, I pushed a fix for the failures. Now it shows that:

```
Predefined type 'System.Range' is not defined or imported
```

seems like a version-compatibility-related problem with C#. I'm a C# beginner, do you know how to resolve it?

**Comment by Leo-LiHao** [comment]
> @Leo-LiHao -- I think this looks really good. I still have two outstanding questions:
> 
> 1. How do you keep buffers and parameters separate when loading the state_dict?
> 2. The Python test doesn't actually seem to validate that the loaded model can be used in both train and eval mode -- shouldn't that be done?

1. I don't distinguish buffers and parameters, would you mind giving an example(or model) that contains both then I can add a test?
2. It can be done. I will find a dataset for doing so (might have to change the model to an "easy" one with fewer parameters)

**Comment by NiklasGustafsson** [comment]
>> would you mind giving an example(or model) that contains both then I can add a test?

BatchNormalization has buffers. Linear (and many others) has parameters. The thing I'm concerned about is that buffers are just regular Tensor instances, while trained weights are 'Parameter' instances. If the code realizes parameters as Tensors, which it looks like it's doing, then can Pytorch deal with that? Does it do something automatic to convert?

Running random data through the model, eval and train mode and have it not generate errors should help there. You don't have to have a real data set, just something of the right shape. Most of our model unit tests just use random data.

**Comment by NiklasGustafsson** [comment]
BTW, this is such a great addition, that I will hold off making another release until we get this in. This, and the NumPy-style ToString() that @dayo05 is working on.

**Comment by Leo-LiHao** [comment]
Hi, @NiklasGustafsson. I added tests for `eval` and `train`. The added test of the LeNet-1 model contains `BatchNorm2d` (has buffers) and `Linear` (has parameters) layers and the code works well. The only thing we should notice is that the names of properties and layers should be the same in C# and Python.

**Review by NiklasGustafsson — APPROVED** [review]

**Comment by NiklasGustafsson** [comment]
I approved the PR. Now, I'm just waiting for the CI results after re-merging with main. If that pans out, I will merge this into main, and then we just need @dayo05's ToString() work to be finished (no pressure :-)) and we are ready for a release.

**Comment by NiklasGustafsson** [comment]
Merged.

@Leo-LiHao, could you also edit the article on saving and loading models to reflect the fact that we now support importing into Python?

**Comment by Leo-LiHao** [comment]
> Merged.
> 
> @Leo-LiHao, could you also edit the article on saving and loading models to reflect the fact that we now support importing into Python?

Thanks. I'm going to do that this weekend.

**Pull request merged**
